rust/src/tests/errors/enotdir.rs

```diff
@@ -0,0 +1,36 @@
+/// Asserts that it returns ENOTDIR if a component of the path prefix is not a directory.
+macro_rules! assert_enotdir_comp {
+    ($syscall: ident, either) => {
+        assert_enotdir_comp!($syscall, |ctx: &mut crate::runner::context::TestContext,
+                                             path: &std::path::Path| {
+                let new_path = ctx.gen_path();
+                let file = path.parent().unwrap();
+                $syscall(path, &new_path).or($syscall(file, path))
+        });
+    };
+
+    ($syscall: ident, $f: expr) => {
+        crate::test_case! {
+            #[doc = concat!(stringify!($syscall),
+                 "returns ENOTDIR if a component of the path prefix is not a directory")]
+            enotdir_component => [Regular, Fifo, Block, Char, Socket]
+        }
+        fn enotdir_component(ctx: &mut crate::runner::context::TestContext,
+                            ft: crate::runner::context::FileType) {
+            let base_path = ctx.create(ft.clone()).unwrap();
+            let path = base_path.join("previous_not_dir");
+
+            assert_eq!($f(ctx, &path).unwrap_err(), nix::errno::Errno::ENOTDIR)
+        }
+    };
+
+    ($syscall: ident $( ($( $($before:expr),* ,)? ~path $(, $($after:expr),*)?) )?) => {
```

Comment:
I find this syntax to be pretty weird, and it doesn't seem to save much code.  I think it would be clearer to remove this macro form, and write
```rust
assert_enotdir_comp!(mkdir, |_ctx, path| mkdir(path, Mode::empty()));
```
instead of
```rust
assert_enotdir_comp!(mkdir(~path, Mode::empty()));
```

Reply:
After spending some more time reading your PRs, I'm getting more comfortable with this unique syntax.  I think it's ok to leave it here, but please add some documentation to the macro explaining what it means.  Doc tests would be useful too.